be/src/formats/parquet/encoding_dict.h

```diff
@@ -203,22 +205,93 @@ class DictDecoder<Slice> final : public Decoder {
     }
 
-    Status get_dict_values(const std::vector<int32_t>& dict_codes, Column* column) override {
+    Status get_dict_values(const std::vector<int32_t>& dict_codes, const NullableColumn& nulls,
+                           Column* column) override {
+        const std::vector<uint8_t>& null_data = nulls.immutable_null_column_data();
+        bool has_null = nulls.has_null();
+        bool all_null = false;
+
+        if (has_null) {
+            size_t count = SIMD::count_nonzero(null_data);
+            all_null = (count == null_data.size());
+        }
+
+        // dict codes size and column size HAVE TO BE EXACTLY SAME.
         std::vector<Slice> slices(dict_codes.size());
-        for (size_t i = 0; i < dict_codes.size(); i++) {
-            slices[i] = _dict[dict_codes[i]];
+        if (!has_null) {
+            for (size_t i = 0; i < dict_codes.size(); i++) {
+                slices[i] = _dict[dict_codes[i]];
+            }
+        } else if (!all_null) {
+            size_t size = dict_codes.size();
+            const uint8_t* null_data_ptr = null_data.data();
+
+            size_t i = 0;
+            for (i = 0; (i + 8) < size; i += 8) {
+                uint64_t partial_all_null;
+                memcpy(&partial_all_null, null_data_ptr, 8);
+                if (partial_all_null != 0) {
+                    for (int j = 0; j < 8; j++) {
+                        if (!null_data_ptr[j]) {
+                            slices[i + j] = _dict[dict_codes[i + j]];
+                        }
+                    }
+                }
+                null_data_ptr += 8;
+            }
+
+            while (i < size) {
+                if (!null_data_ptr[i]) {
+                    slices[i] = _dict[dict_codes[i]];
+                }
+                i += 1;
+            }
+        }
+
+        // if there is no null, we already add tailing bytes to each dict value.
+        // so it's safe to call `append_strings_overflow`
+        // otherwise, slices[i] could be (data=nullptr, size =0), so it's not safe to call that
+        bool ret = false;
+        if (!has_null) {
+            ret = column->append_strings_overflow(slices, _max_value_length);
+        } else {
+            ret = column->append_strings(slices);
         }
-        auto ret = column->append_strings_overflow(slices, _max_value_length);
+
         if (UNLIKELY(!ret)) {
             return Status::InternalError("DictDecoder append strings to column failed");
         }
         return Status::OK();
     }
 
-    Status get_dict_codes(const std::vector<Slice>& dict_values, std::vector<int32_t>* dict_codes) override {
-        for (auto& dict_value : dict_values) {
-            // dict value always exists in _dict_code_by_value
-            dict_codes->emplace_back(_dict_code_by_value[dict_value]);
+    Status get_dict_codes(const std::vector<Slice>& dict_values, const NullableColumn& nulls,
+                          std::vector<int32_t>* dict_codes) override {
```

Comment:
I think we don't need nulls, because after chunk eval conjuncts we just get binarycolumn+null or binary column, we can deal null there and don't need judge on every nulls element here.

Reply:
what's your suggestion? you can not tell null from `dict_values`.